### ISSUE TITLE
No `utest` tests run on Ubuntu x86_64 machine

### ISSUE THREAD

**Issue by mtsokol** [xianyi/OpenBLAS#4815]
Hi!

I built OpenBLAS locally with `make` on two machines: MacOS x86_64 (Haswell), and Ubuntu x86_64.

For each one of them I tried to run OpenBLAS regression tests `utests` with:
```
make CC=cc FC=gfortran -C utest
```

For MacOS it runs all tests:
```
RESULTS: 1526 tests (1526 ok, 0 failed, 0 skipped) ran in 252 ms
```

But for Linux it's empty:
```
RESULTS: 0 tests (0 ok, 0 failed, 0 skipped) ran in 0 ms
```

It looks like no tests are picked in `__ctest_linkTests()` as:
```c
for (test = *(__ctest_head_p); test != NULL; test=test->next) {
    if (test == &__ctest_suite_test) continue;
    // if (filter(test)) total++;
    total++;
}
```
in the `utest/ctest.h` gives total of 0.

Do you know what might be the issue? How to troubleshoot it?

(I see that in the CI Ubuntu+make runs `utest` tests)

**Comment by martin-frbg** [comment]
Strange, what is you C compiler please ? (`cc -v`) So far we know only of the NVidia and Oracle/Sun compilers not picking up the test entry points (and needing to link utest_main2.o instead)

**Comment by mtsokol** [comment]
Here's my output of `cc -v`:
```
Reading specs from /home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/specs
could not find specs file conda.specs
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/home/mtsokol/.conda/envs/openblas-dev/bin/../libexec/gcc/x86_64-conda-linux-gnu/12.3.0/lto-wrapper
Target: x86_64-conda-linux-gnu
Configured with: ../configure --prefix=/home/conda/feedstock_root/build_artifacts/gcc_compilers_1715013812730/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho --with-slibdir=/home/conda/feedstock_root/build_artifacts/gcc_compilers_1715013812730/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib --libdir=/home/conda/feedstock_root/build_artifacts/gcc_compilers_1715013812730/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib --mandir=/home/conda/feedstock_root/build_artifacts/gcc_compilers_1715013812730/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/man --build=x86_64-conda-linux-gnu --host=x86_64-conda-linux-gnu --target=x86_64-conda-linux-gnu --enable-default-pie --enable-languages=c,c++,fortran,objc,obj-c++ --enable-__cxa_atexit --disable-libmudflap --enable-libgomp --disable-libssp --enable-libquadmath --enable-libquadmath-support --enable-libsanitizer --enable-lto --enable-threads=posix --enable-target-optspace --enable-plugin --enable-gold --disable-nls --disable-bootstrap --disable-multilib --enable-long-long --with-sysroot=/home/conda/feedstock_root/build_artifacts/gcc_compilers_1715013812730/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/x86_64-conda-linux-gnu/sysroot --with-build-sysroot=/home/conda/feedstock_root/build_artifacts/gcc_compilers_1715013812730/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/x86_64-conda-linux-gnu/sysroot --with-gxx-include-dir=/home/conda/feedstock_root/build_artifacts/gcc_compilers_1715013812730/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/x86_64-conda-linux-gnu/include/c++/12.3.0 --with-pkgversion='conda-forge gcc 12.3.0-7' --with-bugurl=https://github.com/conda-forge/ctng-compilers-feedstock/issues/new/choose
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 12.3.0 (conda-forge gcc 12.3.0-7)
```

**Comment by martin-frbg** [comment]
Thanks - so it is a normal gcc, apart from being a conda build instead of the stock Ubuntu package. Very curious. What does the generated `Makefile.conf` look like ? (And is the Ubuntu host a regular machine or WSL2 on a Windows host ?)

**Comment by mtsokol** [comment]
Here's `Makefile.conf`:
```properties
OSNAME=Linux
ARCH=x86_64
C_COMPILER=GCC
BINARY32=
BINARY64=1
CEXTRALIB= -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0 -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/lib/../lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../x86_64-conda-linux-gnu/sysroot/lib/../lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../x86_64-conda-linux-gnu/sysroot/usr/lib/../lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../.. -L/home/mtsokol/.conda/envs/openblas-dev/bin/../x86_64-conda-linux-gnu/sysroot/lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../x86_64-conda-linux-gnu/sysroot/usr/lib  -lc 
F_COMPILER=GFORTRAN
FC=gfortran
BU=_
FEXTRALIB= -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0 -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/lib/../lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../x86_64-conda-linux-gnu/sysroot/lib/../lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../x86_64-conda-linux-gnu/sysroot/usr/lib/../lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../.. -L/home/mtsokol/.conda/envs/openblas-dev/bin/../x86_64-conda-linux-gnu/sysroot/lib -L/home/mtsokol/.conda/envs/openblas-dev/bin/../x86_64-conda-linux-gnu/sysroot/usr/lib  -lgfortran -lm -lquadmath -lm -lc 
CORE=ZEN
LIBCORE=zen
NUM_CORES=64
HAVE_MMX=1
HAVE_SSE=1
HAVE_SSE2=1
HAVE_SSE3=1
HAVE_SSSE3=1
HAVE_SSE4_1=1
HAVE_SSE4_2=1
HAVE_SSE4A=1
HAVE_AVX=1
HAVE_AVX2=1
HAVE_FMA3=1
MAKEFLAGS += -j 64
SBGEMM_UNROLL_M=8
SBGEMM_UNROLL_N=4
SGEMM_UNROLL_M=8
SGEMM_UNROLL_N=4
DGEMM_UNROLL_M=4
DGEMM_UNROLL_N=8
QGEMM_UNROLL_M=2
QGEMM_UNROLL_N=2
CGEMM_UNROLL_M=8
CGEMM_UNROLL_N=2
ZGEMM_UNROLL_M=4
ZGEMM_UNROLL_N=2
XGEMM_UNROLL_M=1
XGEMM_UNROLL_N=1
CGEMM3M_UNROLL_M=8
CGEMM3M_UNROLL_N=4
ZGEMM3M_UNROLL_M=4
ZGEMM3M_UNROLL_N=4
XGEMM3M_UNROLL_M=2
XGEMM3M_UNROLL_N=2
```
It's regular Ubuntu host:
```
(openblas-dev) mtsokol@qgpu3:~/OpenBLAS$ uname -a
Linux qgpu3 5.15.0-97-generic #107-Ubuntu SMP Wed Feb 7 13:26:48 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by mtsokol** [comment]
It's the only test suite that isn't executed. `ctest` and `test` both run. Also Adding `OBJS = utest_main2.o` outside of `ifeq` in `utest/Makefile` picks 33 tests.

**Comment by martin-frbg** [comment]
Makefile.conf looks completely normal, no indication that it misidentified the compiler or assumed you were cross-compiling... the 33 tests are correct for openblas_utest, the 1526 are in the separate openblas_utest_ext that tests the BLAS extensions (and probably wants `OBJS_EXT = $(DIR_EXT)/utest_main2.o` too... for whatever reason)

**Comment by mtsokol** [comment]
Hi! my last message was misleading. I wanted to avoid any code changes and make it run without using `utest_main2`. It was just to show that any test can run forcibly.

At the end I managed to get all `utest` tests running the standard way,  with `make` (including `test_extensions`).

My previous setup had `gcc` installed from conda forge with:
```
conda install conda-forge::compilers
```
where the gcc version was `12.3.0`.

I ended up with a new conda env installing:
```
conda install conda-forge::gfortran
```
where the gcc version is `13.2.0`.

With the second setup the tests are picked and run correctly:
```
make -C utest
...
RESULTS: 1526 tests (1526 ok, 0 failed, 0 skipped) ran in 220 ms
```
